### PR TITLE
add type annotations to public routines

### LINKED ISSUE #447: add type hints to public API functions
once we're py3-only, we should consider adding type hints to the "public" API functions to better document how they should be used.

i don't think we should strive to type hint every function (although adding them as we go is fine). doing the public API is a good balance for being helpful without too much work.

### PR DESCRIPTION
**Pull request by williballenthin**
closes #447

invoke like: `mypy --config-file .github/mypy/mypy.ini capa/main.py`

### todo

  - [x] add annotations to `capa.main`
  - [x] add additional lib ignores to mypy.ini
  - [x] fix any type violations identified by mypy
  - [x] configure CI to invoke mypy

### Checklist

- [ ] No CHANGELOG update needed

- [x] No new tests needed

- [x] No documentation update needed

### CODE REVIEW

**Comment by williballenthin** [comment]
generate stubs via runtime tracing using [MonkeyType](https://monkeytype.readthedocs.io/en/latest/index.html)

```
pip install MonkeyType
monkeytype run capa/main.py tests/data/Practical\ Malware\ Analysis\ Lab\ 01-01.dll_
```
(this takes much longer than without tracing, matching takes around 1.5s/function)

distribution of trace data:
image
(incidentally, this shows what our hot functions are)

**Comment by williballenthin** [comment]
- [ ] capa.engine
  - [ ] capa.features
  - [ ] capa.features.basicblock
  - [ ] capa.features.extractors
  - [ ] capa.features.extractors.helpers
  - [ ] capa.features.extractors.loops
  - [ ] capa.features.extractors.pefile
  - [ ] capa.features.extractors.strings
  - [ ] capa.features.extractors.viv
  - [ ] capa.features.extractors.viv.basicblock
  - [ ] capa.features.extractors.viv.file
  - [ ] capa.features.extractors.viv.function
  - [ ] capa.features.extractors.viv.helpers
  - [ ] capa.features.extractors.viv.indirect_calls
  - [ ] capa.features.extractors.viv.insn
  - [ ] capa.features.file
  - [ ] capa.features.freeze
  - [ ] capa.features.insn
  - [ ] capa.helpers
  - [ ] capa.render
  - [ ] capa.render.default
  - [ ] capa.render.utils
  - [ ] capa.rules

**Comment by williballenthin** [comment]
blocked on #622 right now

**Comment by williballenthin** [comment]
my convention, add type annotations when:
  - the function is "public"
  - the function is reused often internally, so mypy could catch a misuse
  - the annotations don't add much noise, i.e. they dont look like `Tuple[Dict[str, Set[.....`
    - don't use annotations for nested dicts, rely on docstrings to explain/example

basically, don't bend over backwards for type annotations, but add them where they make sense. hopefully we all have the same sense of style, but if not, i'm happy to work with you all to find a happy balance.

**Comment by williballenthin** [comment]
wait for merge of #622 before reviewing this so the diff collapses.

**Comment by williballenthin** [comment]
current results:

```
> mypy --config-file .github/mypy/mypy.ini capa/                                                                                 feature-447!?
capa/features/common.py:217: error: Argument 1 to "__init__" of "Regex" has incompatible type "Union[str, int, bytes]"; expected "str"
capa/rules.py:21: error: Name "lru_cache" already defined (possibly by an import)
capa/rules.py:147: error: Argument 1 to "Characteristic" has incompatible type "Union[str, int, bytes]"; expected "str"
capa/rules.py:149: error: Argument 2 to "filter" has incompatible type "Set[object]"; expected "Iterable[Union[type, Tuple[Any, ...]]]"
capa/rules.py:178: error: Incompatible types in assignment (expression has type "int", variable has type "str")
capa/rules.py:179: error: Unsupported operand types for < ("str" and "int")
capa/rules.py:182: error: Incompatible types in assignment (expression has type "None", variable has type "str")
capa/rules.py:185: error: Incompatible types in assignment (expression has type "int", variable has type "str")
capa/rules.py:186: error: Unsupported operand types for < ("str" and "int")
capa/rules.py:189: error: Incompatible types in assignment (expression has type "None", variable has type "str")
capa/rules.py:260: error: Incompatible types in assignment (expression has type "Union[str, int]", variable has type "str")
capa/rules.py:265: error: Incompatible types in assignment (expression has type "bytes", variable has type "str")
capa/rules.py:275: error: Incompatible types in assignment (expression has type "int", variable has type "str")
capa/render/default.py:134: error: Argument 1 to "add" of "set" has incompatible type "Tuple[Any, Any]"; expected "Tuple[Any, Any, Any]"
capa/render/default.py:141: error: Too many values to unpack (2 expected, 3 provided)
capa/render/default.py:142: error: Cannot determine type of "technique"
capa/render/default.py:142: error: Cannot determine type of "id"
capa/render/default.py:187: error: Argument 1 to "add" of "set" has incompatible type "Tuple[Any, Any]"; expected "Tuple[Any, Any, Any]"
capa/render/default.py:194: error: Too many values to unpack (2 expected, 3 provided)
capa/render/default.py:195: error: Cannot determine type of "behavior"
capa/render/default.py:195: error: Cannot determine type of "id"
capa/features/extractors/viv/indirect_calls.py:164: error: Incompatible return value type (got "Tuple[int, Optional[int]]", expected "Tuple[int, int]")
Found 22 errors in 4 files (checked 57 source files)
```

**Review by mr-tz — COMMENTED** [review]
skimmed this as well, looks good so far

File: capa/engine.py [hunk 1]
```diff
@@ -33,7 +45,7 @@ def __str__(self):
     def __repr__(self):
         return str(self)
 
-    def evaluate(self, ctx):
+    def evaluate(self, features: FeatureSet) -> "Result":
```
Comment (williballenthin) [inline capa/engine.py:48]: we have a circular dep Statement -> Result -> Statement, so we have to use this silly syntax with quotes to defer the resolution of the type.

typically i'd just leave off the type hint, but this is a key routine, so we really want hints here.

**Review by mr-tz — APPROVED** [review]
wow, I skimmed over this and noticed great improvements in addition to the type hints
maybe another reviewer can provide some feedback as well, but I'm also good with merging and testing more intensely as we go towards 2.0 in the next few weeks.

**Comment by williballenthin** [comment]
@mike-hunhoff can you take a peek today?

**Review by mike-hunhoff — APPROVED** [review]
Thank you for all of the work @williballenthin! I'm with @mr-tz it looks good; let's get it merged.

**Pull request merged**